File: BilisPera/BilisPera/Categories/UIViewController+Extension.m
//
//  UIViewController+Extension.m
//  BilisPera
//
//  Created by BHJ on 2025/8/11.
//

#import "UIViewController+Extension.h"

@implementation UIViewController (Extension)

+ (UIViewController *)topMost {
    NSArray<UIWindow *> *currentWindows = [self getCurrentWindows];
    if (!currentWindows) {
        return nil;
    }
    UIViewController *rootViewController = [self getRootViewController:currentWindows];
    if (!rootViewController) {
        return nil;
    }
    return [self topMostOf:rootViewController];
}

+ (UIApplication *)sharedApplication {
    return [UIApplication sharedApplication];
}

+ (NSArray<UIWindow *> *)getCurrentWindows {
    return [[self sharedApplication] windows];
}

+ (UIViewController *)getRootViewController:(NSArray<UIWindow *> *)currentWindows {
    for (UIWindow *window in currentWindows) {
        if (window.isKeyWindow) {
            return window.rootViewController;
        }
    }
    return nil;
}
+ (UIViewController *)topMostOf:(UIViewController *)viewController {
    // presentedViewController
    if (viewController.presentedViewController) {
        return [self topMostOf:viewController.presentedViewController];
    }
    
    // tabBarController
    if ([viewController isKindOfClass:[UITabBarController class]]) {
        UITabBarController *tabBarController = (UITabBarController *)viewController;
        if (tabBarController.selectedViewController) {
            return [self topMostOf:tabBarController.selectedViewController];
        }
    }
    
    // UINavigationController
    if ([viewController isKindOfClass:[UINavigationController class]]) {
        UINavigationController *navigationController = (UINavigationController *)viewController;
        if (navigationController.visibleViewController) {
            return [self topMostOf:navigationController.visibleViewController];
        }
    }
    
    // UIPageViewController
    if ([viewController isKindOfClass:[UIPageViewController class]]) {
        UIPageViewController *pageViewController = (UIPageViewController *)viewController;
        if (pageViewController.viewControllers.firstObject) {
            return [self topMostOf:pageViewController.viewControllers.firstObject];
        }
    }
    
    // childViewController
    for (UIView *subView in viewController.view.subviews) {
        if ([subView.nextResponder isKindOfClass:[UIViewController class]]) {
            return [self topMostOf:(UIViewController *)subView.nextResponder];
        }
    }
    return viewController;
}

+ (UINavigationController *)topMostNavigationControllerOf:(UIViewController *)viewController {
    // presentedViewController
    if (viewController.presentedViewController) {
        return [self topMostNavigationControllerOf:viewController.presentedViewController];
    }
    
    // tabBarController
    if ([viewController isKindOfClass:[UITabBarController class]]) {
        UITabBarController *tabBarController = (UITabBarController *)viewController;
        if (tabBarController.selectedViewController) {
            return [self topMostNavigationControllerOf:tabBarController.selectedViewController];
        }
    }
    
    // UINavigationController
    if ([viewController isKindOfClass:[UINavigationController class]]) {
        UINavigationController *navigationController = (UINavigationController *)viewController;
        if (navigationController.visibleViewController) {
            return [self topMostNavigationControllerOf:navigationController.visibleViewController];
        }
    }
    return viewController.navigationController;
}

-(void)popNavigation:(BOOL)animated{
    if (self.presentingViewController) {
        [self dismissViewControllerAnimated:animated completion:nil];
    }else{
        [self.navigationController popViewControllerAnimated:animated];
    }
}

-(void)presentFullScreen{
    self.modalPresentationStyle = UIModalPresentationOverCurrentContext;
    [self setModalInPresentation:YES];
}

-(void)showCustomAlertViewWith:(BPAlertViewModel *)model{
    BPAlertViewController *alertVC = [[BPAlertViewController alloc] initWith:model];
    [alertVC presentFullScreen];
    [self presentViewController:alertVC animated:NO completion:nil];
}

-(void)dismisCustomAlertView:(nullable simpleCompletion)completion{
    UIViewController *topVC = [UIViewController topMost];
    if ([topVC isKindOfClass:[BPAlertViewController class]]) {
        [topVC dismissViewControllerAnimated:false completion:completion];
    }
}

- (void)showCustomAlertWithTitle:(NSString *)title
                         message:(NSString *)message
               confirmCompletion:(simpleCompletion)confirmCompletion
                cancelCompletion:(simpleCompletion)cancelCompletion {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                             message:message
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    // 设置警告框视图背景色和文字颜色
    alertController.view.backgroundColor = [UIColor whiteColor];
    // 设置消息文字颜色和字体
    if ([alertController.view.subviews.firstObject.subviews.firstObject isKindOfClass:[UILabel class]]) {
        UILabel *label = (UILabel *)alertController.view.subviews.firstObject.subviews.firstObject;
        label.textColor = kBlackColor;
        label.font = kFont(14);
    }
    // 确认按钮
    UIAlertAction *confirmAction = [UIAlertAction actionWithTitle:@"ok"
                                                            style:UIAlertActionStyleDestructive
                                                          handler:^(UIAlertAction * _Nonnull action) {
        if (confirmCompletion) {
            confirmCompletion();
        }
    }];
    [confirmAction setValue:[UIColor blueColor] forKey:@"titleTextColor"];
    [alertController addAction:confirmAction];
    
    // 取消按钮
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:^(UIAlertAction * _Nonnull action) {
        [self dismissViewControllerAnimated:YES completion:nil];
        if (cancelCompletion) {
            cancelCompletion();
        }
    }];
    [cancelAction setValue:[UIColor grayColor] forKey:@"titleTextColor"];
    [alertController addAction:cancelAction];
    // 显示警告框
    [self presentViewController:alertController animated:YES completion:nil];
}



@end

File: SkyLoan_phillipines/Extension/UIViewController+Extension.swift
//
//  UIViewController+Extension.swift
//  SkyLoan_phillipines
//
//  Created by BHJ on 2025/5/14.
//

import UIKit

protocol TopMostViewControllerProviding{
    var topMostViewController: UIViewController? {get}
}

struct TopMostViewControllerProvider: TopMostViewControllerProviding {
    var topMostViewController: UIViewController?{
        UIViewController.topMost
    }
}

extension UIViewController{
    @objc func presentFullScreenAndDisablePullToDismiss(){
        modalPresentationStyle = .overCurrentContext
        if #available(iOS 13.0, *){
            isModalInPresentation = true
        }
    }
    
    func showCustomAlert(model: CustomAlertView.Model = .init()){
        let alertVC = CustomAlertView.init(model: model)
        present(alertVC, animated: false)
    }
    
    func hideCustomAlertView(completion: (()-> Void)? = nil){
        guard let topVC = UIViewController.topMost as? CustomAlertView else {
            return
        }
        topVC.dismiss(animated: false,completion: completion)
    }
    
   @objc func popNavigation(animated: Bool = true){
        if let _ = self.presentingViewController {
            dismiss(animated: animated)
        }else{
            navigationController?.popViewController(animated: animated)
        }
    }
    
    func showCustomAlert(title: String,message: String,confirmCompletion: (()-> Void)? = nil,cancelCompletion: (()-> Void)? = nil) {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let alertWindow = alertController.view
        alertWindow?.backgroundColor = .white
        alertWindow?.tintColor = kColor_black
        if let label = alertController.view.subviews.first?.subviews.first as? UILabel {
            label.textColor = UIColor.black
            label.font = SLFont(size: 14)
        }
        let confirmAction = UIAlertAction(title: "confirm", style: .destructive) { action in
            confirmCompletion?()
        }
        confirmAction.setValue(UIColor.blue, forKeyPath: "_titleTextColor")
        alertController.addAction(confirmAction)
        let cancelAction = UIAlertAction(title: "cancel", style: .cancel) { action in
            self.dismiss(animated: true)
            cancelCompletion?()
        }
        cancelAction.setValue(UIColor.gray, forKeyPath: "_titleTextColor")
        alertController.addAction(cancelAction)
        // 显示警告框
        self.present(alertController, animated: true, completion: nil)
    }
}

extension UIViewController{
    
    @objc public class var topMost: UIViewController?{
        guard let currentWindows = getCurrentWindows() else {
            return nil
        }
        guard let rootViewController = getRootViewController(currentWindows: currentWindows) else {
            return nil
        }
        return topMost(of: rootViewController)
    }
    
    public class var sharedApplication: UIApplication?{
        let selector = NSSelectorFromString("sharedApplication")
        return UIApplication.perform(selector).takeUnretainedValue() as? UIApplication
    }
    
    class func getCurrentWindows() -> [UIWindow]?{
        sharedApplication?.windows
    }
    
    class func getRootViewController(currentWindows: [UIWindow]) -> UIViewController?{
        currentWindows.filter({$0.isKeyWindow}).first?.rootViewController
    }
    
    public class func topMost(of viewController: UIViewController) -> UIViewController{
        // presentedViewController
        if let presentedViewController = viewController.presentedViewController {
            return topMost(of: presentedViewController)
        }
        // tabBarController
        if let tabBarController = viewController as? UITabBarController,let selectedViewController = tabBarController.selectedViewController{
            return topMost(of: selectedViewController)
        }
        // UINavigationController
        if let navigationController = viewController as? UINavigationController,let visibedViewController = navigationController.visibleViewController{
            return topMost(of: visibedViewController)
        }
        // UIPageViewController
        if let pageViewController = viewController as? UIPageViewController,let viewController = pageViewController.viewControllers?.first{
            return topMost(of: viewController)
        }
        // childViewController
        for subView in viewController.view.subviews{
            if let childViewController = subView.next as? UIViewController{
                return topMost(of: childViewController)
            }
        }
        return viewController
    }
    
    public class func topMostNavigationController(of viewController: UIViewController) -> UINavigationController?{
        // presentedViewController
        if let presentedViewController = viewController.presentedViewController {
            return topMostNavigationController(of: presentedViewController)
        }
        // tabBarController
        if let tabBarController = viewController as? UITabBarController,let selectedViewController = tabBarController.selectedViewController{
            return topMostNavigationController(of: selectedViewController)
        }
        // UINavigationController
        if let navigationController = viewController as? UINavigationController,let visibedViewController = navigationController.visibleViewController{
            return topMostNavigationController(of: visibedViewController)
        }
        return viewController.navigationController
    }
    
    @objc public class var topMostTabBarController: UITabBarController?{
        guard let currentWindows = getCurrentWindows() else {
            return nil
        }
        return getTopMostTabBarController(currentWindows: currentWindows)
    }
    
    class func getTopMostTabBarController(currentWindows: [UIWindow]) -> UITabBarController?{
        guard let rootViewController = getRootViewController(currentWindows: currentWindows) else {
            return nil
        }
        return topMostTabBarController(of: rootViewController)
    }
    
    public class func topMostTabBarController(of controller: UIViewController) -> UITabBarController?{
        if let result = controller as? UITabBarController{
            return result
        }
        if let result = controller.tabBarController{
            return result
        }
        // presentedViewController
        if let presentedViewController = controller.presentedViewController{
            return topMostTabBarController(of: presentedViewController)
        }
        // childViewController
        for subView in controller.view.subviews{
            if let childViewController = subView.next as? UIViewController{
                return topMostTabBarController(of: childViewController)
            }
        }
        return nil
    }
}


extension UIScene.ActivationState{
    var sortPriority: Int{
        switch self {
        case .unattached:
            return 4
        case .foregroundActive:
            return 1
        case .foregroundInactive:
            return 2
        case .background:
            return 3
        @unknown default:
            return 5
        }
    }
}

public extension UIViewController{
    private static var animationTypeKey = "animationTypeKey";
    var animationType:UIViewControllerAnimationType{
        set {
            objc_setAssociatedObject(self,  &UIViewController.animationTypeKey, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        }
        get {
            if objc_getAssociatedObject(self, &UIViewController.animationTypeKey) is UIViewControllerAnimationType {
                return (objc_getAssociatedObject(self, &UIViewController.animationTypeKey) as? UIViewControllerAnimationType)!;
            }
            return UIViewControllerAnimationType(rawValue: 0)!;
        }
    }
}
